File: Python/백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque

N = int(input())
map_ = [list(map(int, input().split())) for _ in range(N)]
visited = [[False] * N for _ in range(N)]
ans = 1
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]

def BFS(x, y, h):
    q = deque()
    q.append((x, y))
    visited[y][x] = True
    
    while q:
        curX, curY = q.popleft()
        for i in range(4):
            nxtX = curX + dx[i]
            nxtY = curY + dy[i]
            if 0 <= nxtX < N and 0 <= nxtY < N:
                if map_[nxtY][nxtX] > h and not visited[nxtY][nxtX]:
                    visited[nxtY][nxtX] = True
                    q.append((nxtX, nxtY))

def count_area(h):
    global ans
    for i in range(N):
        for j in range(N):
            visited[i][j] = False
            
    num = 0
    for i in range(N):
        for j in range(N):
            if map_[i][j] > h and not visited[i][j]:
                BFS(j, i, h)
                num += 1
    
    if num > ans:
        ans = num
        
mxH = max(max(row) for row in map_)
for h in range(1, mxH):
    count_area(h)
    
print(ans)

File: C++17/백준/Silver/2468. 안전 영역/안전 영역.cc
#include <iostream>
#include <queue>
using namespace std;
int map[100][100];
bool visited[100][100];
int N;
int ans = 1;
int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
#define X first
#define Y second

void BFS(int x, int y, int h) {
    queue<pair<int, int>> q;
    q.push({x, y});
    visited[y][x] = true;
    
    while(!q.empty()) {
        auto cur = q.front(); q.pop();
        for (int i = 0; i < 4; i++) {
            int nxtX = cur.X + dx[i];
            int nxtY = cur.Y + dy[i];
            if (nxtX < 0 || nxtX >= N || nxtY < 0 || nxtY >= N) continue;
            if (map[nxtY][nxtX] > h && !visited[nxtY][nxtX]) {
                visited[nxtY][nxtX] = true;
                q.push({nxtX, nxtY});
            }
        }
    }
}

void countArea(int h) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            visited[i][j] = false;
        }
    }
    
    int num = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            if (map[i][j] > h && !visited[i][j]) {
                BFS(j, i, h);
                ++num;
            }
        }
    }
    if (num > ans)
        ans = num;
}

int main() {
    cin >> N;
    vector<vector<int>> v(N);
    int mxH = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            cin >> map[i][j];
            if (mxH < map[i][j])
                mxH = map[i][j];
        }
    }
    for (int i = 1; i < mxH; i++) {
        countArea(i);
    }
    cout << ans;
}